### DISCUSSION TITLE
How to recursively clone submodules?

### DISCUSSION THREAD

**Discussion by NobodyXu** [GitoxideLabs/gitoxide#1001] [Q&A] [answered]
In the latest release, I see that there are several new types related to submodule.
I read the `Submodule` struct, however I'm not sure how to recursively clone it.

Is it not supported yet?

**Comment by Byron** [comment, marked as answer]
There are a couple of operations that can be applied to submodules like 'init' or 'update' which all are not implemented yet. However, all the primitives are there to support implementing it by hand. For instance, I plan to add `--recurse-submodules` for `gix clone`, which can initially be in `gitoxide-core` and then move to something more general into `gix` from there.

**Reply by NobodyXu** [reply]
Thanks, I'm looking forward to it!
Hopefully `gix` would also support `recurse_submodules`.

**Reply by Byron** [reply]
If `reset()` deals with submodules, the `gix` will as well. The idea is to honor the various configuration variables that control how submodules are handled when fetching, for example, I just don't have an understanding yet of how that all would look like. As it's not needed for `cargo` though it will be a while until that is implemented though. Thus for submodule handling has to be done by hand.

**Reply by koppor** [reply]
In a [discussion of on-demand installation of a git command supporting clone only](https://github.com/eirikb/gg/discussions/88#discussioncomment-13244090), I found `gitoxide`. Since I need `--recurse-submodules`, I found this thread.

May I ask if there was progress? 😅

On Windows, I get

```cmd
>C:\Users\koppor\.cargo\bin\gix.exe clone --recurse-submodules https://github.com/JabRef/jabref.git JabRef
error: unexpected argument '--recurse-submodules' found

  tip: to pass '--recurse-submodules' as a value, use '-- --recurse-submodules'

Usage: gix.exe clone [OPTIONS] <REMOTE> [DIRECTORY]

For more information, try '--help'.
```

(I'm using ` gitoxide v0.44.0 (x86_64-pc-windows-msvc)` installed via `cargo binstall gitoxide`)

**Reply by Byron** [reply]
This wasn't implemented yet, even though submodule handling does exist to enable it.

There is probably many details involved to get this right, so I'd expect this to be more involved than a simple recursive loop like one would initially assume.

**Reply by koppor** [reply]
I think, later the complexity comes from handling [`.gitmodules`](https://git-scm.com/docs/gitmodules) right.

**Reply by Byron** [reply]
This may be just me who assumes there will be unanticipated details, to stay vigilant during the implementation. Much seems simple at first, but usually that's due to ignorance. And right now, I am pretty ignorant as I didn't research this yet.
